### DISCUSSION TITLE
Idiomatic way to nest object inside of each other

### DISCUSSION THREAD

**Discussion by Zekfad** [rrousselGit/riverpod#1514] [Q&A] [answered]
What is the idiomatic way to create object (1) with nested object (2), so you can listen on both of them?

More context:
- Suppose you have a page with **list of books**, each book has **views count** and an **author**.
- You can open **book page** and therefore **views count increments** (via API).
- Each book is a model **class** returned by API, so provider for book in list and book on book page is the same.

Additionally:
- Each book has **author** with **total author's books view count**.
- Similarly to books author has page and so on.

In total we have:
- Book object with updatable counter and author object.
- Author object with updatable counter and books list.

We want to be able to individually listen to books and authors while they depend on each other.

The main problem here is they we don't want to place providers inside of data model classes, and even doing so isn't clear what to do.

**Comment by rrousselGit** [comment, marked as answer]
The trick is to not store the object itself in the model, but rather its ID.

You'd typically have two separate providers. One for authors, one for books.

At the same time, Authors shouldn't have their list of books. Instead they should have the IDs of their books.
Same goes for books. They shouldn't have their authors. But rather the ID of their authors:

```dart
class Book {
  final String id;
  final String title;
  final List<String> authorIds;
  final int views;
}

class Author {
  final String id;
  final String name;
  final List<String> bookIds;
}
```


Then, the two providers could be something  like:

```dart
final bookProvider = FutureProvider.family.autoDispsoe<Book, String>((ref, id) async {
   return Book.fromJson(await httpGet('my-api/books/$id'));
});

final authorProvider = FutureProvider.family.autoDispsoe<Author, String>((ref, id) async {
   return Author.fromJson(await httpGet('my-api/authors/$id'));
});
```


You can then get the list of books for an author with:

```dart
Widget build(context, ref) {
  final author = ref.watch(authorProvider('some id'));
  return author.when(
    // TODO handle loading/error
   data: (author) {
     return ListView.builder(
       itemCount: author.bookIds.length,
       itemBuilder: (context, index) {
         final bookId = author.bookIds[index];
         final book = ref.watch(bookProvider(bookId));
        return BookView(book: book);
       },
     );
    }
  );
}
```

This logic could be reversed to get the authors from a book:

```dart
Widget build(context, ref) {
  final book = ref.watch(bookProvider('some id'));
  return book.when(
    // TODO handle loading/error
   data: (book) {
     return ListView.builder(
       itemCount: book.authorIds.length,
       itemBuilder: (context, index) {
         final authorId = book.authorIds[index];
         final author = ref.watch(authorProvider(authorId));
        return AuthorView(author: author);
       },
     );
    }
  );
}
```


For the total views count of an author, if you were to do that client side, you could do:

```dart
final totalViewsCountForAuthorProvider = FutureProvider.autoDispose.family<int, String>((ref, authorId) async {
  final author = await ref.watch(authorProvider(authorId).future); // we get the author

  var totalViewsCount = 0;
  // TODO use `Future.wait` to query all books in parallel instead of sequential
  for (final bookId in author.bookIds) {
    final book = await ref.watch(bookProvider(bookId));
    totalViewsCount += book.views;
  }
  return totalViewsCount;
});
```



From there, updating the view count of a book could be a matter of doing:

```dart
Button(
  onPressed: () async {
    await httpPost('api/read?book=$bookId');
    ref.invalidate(bookProvider(bookId)); // The server should have updated the viewCount. Let's re-fetch the book
  }
)
```

The `totalViewsCount` for the associated authors would automatically be recomputed.

**Reply by Zekfad** [reply]
Thank you very much for your answer!

Regarding `ListView.builder` from your snippets, shouldn't this be localized? I mean in current state reload of a book in inside of the list will cause rebuild of the whole list instead of rebuilding single item in it? Or maybe we should pass id to item and subscribe to it from within `BookView` in your example?
What do you think is the better solution?

**Reply by rrousselGit** [reply]
Optimizing rebuilds is a different topic.

There are various options, such as `select` or splitting your UI into multiple `const` widgets.

**Reply by Zekfad** [reply]
Thanks for reply!

**Comment by scopendo** [comment]
I've seen variants of this Riverpod discussion before and have never been able to get myself onboard with the suggested approach.  I can see how it would work but I would be very concerned about performance, making a separate API call for each id.

For example, suppose this fictitious app allows you to tap on an author and see a screen with all their books.  Ignoring pagination, I'd want to be able to make a single API call that returns all the required book data for the author, not just the ids.

I'd be on board with the suggested approach if there was a way to tell Riverpod "here's the value for a provider, which I've fetched separately".  It could be done with provider scope overrides on a family but that's really messy, imo.

**Reply by rrousselGit** [reply]
Riverpod will probably contain something to dedup requests at some point.

But you could build one on your own. I don't think that's really a big deal

**Reply by scopendo** [reply]
I did consider that at one point but it felt like I was reimplementing a secondary caching layer that wasn't easily synchronised with the application and was trying to replicate a subset of what Riverpod offers.

**Reply by rrousselGit** [reply]
I agree.
Although technically speaking, that's what people use the Repository pattern for already. 

I'd like to cover the other feature requests first (offline & co). But that's a problem I'll try to cover.
